### DISCUSSION TITLE
Pi Zero 2 W Install Hangs

### DISCUSSION THREAD

**Discussion by JK-443** [mKeRix/room-assistant#967] [Q&A] [answered]
So I've seen similar posts and have tried every piece of advice I've found so far, but nothing has worked for me. I'm new to HA and Room Assistant, so my apologies in advance for being in over my head here....hoping someone might be able to shed some light on this though!

Each time installing room assistant on my Pi Zero 2 W devices it hangs in exactly the same spot and the install never completed. I've tried following the instructions for the Pi Zero W as well as the one for the Pi 3/4 and, while I did manage a successful installation ONCE with the latter process, I cannot seem to repeat that success.

I've done **sudo apt update** and **sudo apt upgrade** pretty much every time, started from scratch with a new flash dozens of times, tried manually updating some of the dependencies indicated in the WARN errors, but nothing works. 

```
pi@**********:~ $ sudo npm i --global --unsafe-perm room-assistant
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated tar@2.2.2: This version of tar is no longer supported, and will not receive security updates. Please upgrade asap.
⸨##################⸩ ⠼ reify:mathjs: http fetch GET 200 https://registry.npmjs.org/mathjs/-/mathjs-9.5.2.tgz 85657ms (cache miss)
```

It always hangs on that last spot with "reify.mathjs:"...thoughts?

**Comment by PeteBa** [comment, marked as answer]
So, the RP3 instructions are the correct ones as the Zero 2 uses the same chip. The extra "sudo apt update/upgrade" isnt required but shouldnt do any harm. I would recommend against manually updating the dependencies as the older ones have been tested to work with room-assistant whereas the newer one havent (caveat would be if you expose the room-assistant service to the internet). As stated in the room-assistant instructions WARN messages are OK so long as the overall operation completes successfully.

Out of interest, how long do you wait before concluding that the `npm` command has hung ? Reason I ask is that I have had installs stay on that line for up to 35mins before printing the next notice. Can you run the install with `--verbose` and `--foreground-scripts` so that we can see what might be going on. This will produce a lot of output but will also show if thing are continuing on in the background or are really stuck. Start a fresh install but run:

```
sudo npm i --global --unsafe-perm --verbose --foreground-scripts room-assistant
```

**Reply by eLVee1991** [reply]
I am experiencing the same problem as mentioned by @JK-443 . This happens to me too if I try to do this over WiFi and LAN via SSH. This is because of a broken pipe. I am currently running another install following the Raspberry Pi 3/4 guide. NPM is stuck at installing mathjs for about 40 minutes now. If this results in another broken pipe, I will post an update. I'll try your install command mentioned above if this is the case.

**Reply by eLVee1991** [reply]
Another broken pipe after one hour. Will run the installation with command mentioned by @PeteBa and post results later.

**Reply by JK-443** [reply]
@PeteBa, I followed your advice and had similar results to @eLVee1991. Have had one attempt fail citing network issues but have managed to fully install on 3 additional Pis since then. To answer your earlier question, I had let several of the installs work for 1-2 hours with no progress past that one point. I suspect that your assertion about the session timing out after a period of perceived inactivity may be correct.

For anyone interested, I first run` sudo apt update && sudo apt upgrade` as soon as I've managed to SSH in and then follow the room-assistant RP3/4 install instructions, substituting the `sudo npm i --global --unsafe-perm --verbose --foreground-scripts room-assistant` line.

In any event, thanks very much for your help on this! I had spent many hours banging my head against the proverbial wall so, however simple the solution may have been, I'm pretty happy to have it!

Cheers!

**Reply by JmpnJimBob** [reply]
I was having the same “hang” issue as mentioned above with a new install of Room-Assistant on a RPi 3a. 

I took @JK-443 suggestion and performed `sudo apt update && sudo apt upgrade` immediately after I connected with SSH.

I then followed the RPi3 installation instructions and when it was time I once again used @JK-443 substitution for the Room-Assistant install command line `sudo npm i --global --unsafe-perm --verbose --foreground-scripts room-assistant`

Everything went smoothly and Room-Assistant successfully installed on my RPi3a. This was after at least half a dozen failed attempts. Thanks @JK-443

**Reply by gasecki** [reply]
Same here, think that `--verbose --foreground-scripts`opitions do the job. Thanks

**Comment by eLVee1991** [comment]
I ran the command and this seems to finish the install in about 10 mins over LAN. 
`sudo npm i --global --unsafe-perm --verbose --foreground-scripts room-assistant`

Ran the regular command on another pi zero w 2 and got same issue with freezing at installing mathjs.  Can't really tell why. 
`sudo npm i --global --unsafe-perm room-assistant`

**Comment by PeteBa** [comment]
So once npm has downloaded mathjs then it starts to build a bunch of modules that take quite a bit of time. In previous versions of npm, the build scripts would be run in the foreground and the associated output sent to the screen. The newer npm command runs the scripts in the background and swallows the output by default.

Could be that the pi is terminating the ssh session thinking it is inactive as the regular command will go for some time without generating output during the build phase. This will also terminate all processes running in that shell including the npm command. Running with foreground-scripts enabled could be creating enough traffic to avoid this. You could check if your client has an "SSH Keep Alive" option and try setting that ?

When you get into the "hung" situation, are you still able to ping the pi and even ssh back in to see if the npm command is still running ? If the pi is still running but the npm command is not then that would lend weight to a "keep-alive" problem.

**Comment by eLVee1991** [comment]
@PeteBa Thank you for your explanation, makes sense! I've also tested this on an old machine running Rasbian Duster following the pi 3/4 install guide. No freezing with npm on mathjs there. So maybe a pi zero w2 specific problem?

I will check your suggestions about pinging and get back soon. If there is any more testing I can do let me know. Got some spare time in the evenings.

**Reply by eLVee1991** [reply]
I can't SSH back into the machine after freezing. Also not from another client. 

```
ping roomassistant-lr.local
Pinging roomassistant-lr.local [192.168.2.210] with 32 bytes of data:
Reply from 192.168.2.210: bytes=32 time=20ms TTL=64
Reply from 192.168.2.210: bytes=32 time=13ms TTL=64
Reply from 192.168.2.210: bytes=32 time=5ms TTL=64
Reply from 192.168.2.210: bytes=32 time=15ms TTL=64

Ping statistics for 192.168.2.210:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 5ms, Maximum = 20ms, Average = 13ms
```

![roomassistant-lr install freeze](https://user-images.githubusercontent.com/14894071/143046738-b24c88b2-8f38-4341-8abc-0968136dc3f4.png)

**Comment by PeteBa** [comment]
I managed to replicate this on one of my Zero 2 Ws. For what it is worth, looks like a deeper root-cause.

Basically, npm sees that the Z2W has four cores and wants to run a build process on each. But the Z2W only has 512MB of memory so that quickly gets exhausted along with available swap space. The pi then thrashes between the four build processes spending more time swapping out memory than actually progressing the build. Looks like everything grinds to a halt and eventually times out. During this period, the pi is still "operating" (e.g. responding to pings) but there isn't enough spare cycles to deal with a new SSH request. With the `foreground-scripts` enabled then the pi will only start a single build process and that can operate within the 512MB memory.

HTOP with npm in "normal" mode showing multiple build processes and swap space thrashing i.e.  max'ed disk IO:
![image](https://user-images.githubusercontent.com/16738570/143061917-70bb2f8e-2445-4136-96b1-6909d66bf933.png)

HTOP with npm in "foreground-script" mode showing single build process operating just within limits of available memory/swap:
![image](https://user-images.githubusercontent.com/16738570/143061799-a5490544-39bc-455f-8dab-906bb1c6009d.png)

I agree that this looks like a Z2W problem, on a Z1W there is only one core and therefore one build thread that can be managed. Conversely, on a Pi3, there is 1GB of memory to feed the four build threads. The Z2W feels a very lopsided device (wrt cpu vs memory). Be interesting to see how it gets used. I dont see any obvious advantage as a dedicated room-assistant device over the original Zero W.

**Comment by HiTechTLS** [comment]
I was able to solve this by increasing the default swap space from 100 to 1024. The default 100MB of swap is not sufficient and both RAM and swap run out while installing and subsequently hang SSH. Less concerned about runtime, this was more of a problem during install. A long 45 minute install... Might be of value to put either (this or "sudo npm i --global --unsafe-perm --verbose --foreground-scripts room-assistant") as a recommendation on the instructions page for this, as configuration prior to install: https://www.room-assistant.io/guide/quickstart-pi-zero-w.html
